pkg/registry/common/expire/ns_server.go

```diff
@@ -30,88 +29,135 @@ import (
 	"github.com/networkservicemesh/api/pkg/api/registry"
 )
 
-type nsServer struct {
+type expireNSServer struct {
 	nseClient  registry.NetworkServiceEndpointRegistryClient
 	monitorErr error
-	timers     timerMap
-	nsCounts   intMap
-	contexts   contextMap
 	once       sync.Once
 	chainCtx   context.Context
+	nsStates   nsStateMap
 }
 
-func (n *nsServer) checkUpdates() {
-	c, err := n.nseClient.Find(n.chainCtx, &registry.NetworkServiceEndpointQuery{
-		NetworkServiceEndpoint: &registry.NetworkServiceEndpoint{},
-		Watch:                  true,
-	})
-	if err != nil {
-		n.monitorErr = err
-		return
-	}
-	for nse := range registry.ReadNetworkServiceEndpointChannel(c) {
-		for i := range nse.NetworkServiceNames {
+type nsState struct {
+	Timers  map[string]*time.Timer
+	Context context.Context
+	sync.Mutex
+}
+
+func (n *expireNSServer) checkUpdates(eventCh <-chan *registry.NetworkServiceEndpoint) {
+	for event := range eventCh {
+		nse := event
+		if nse.ExpirationTime == nil {
+			continue
+		}
+
+		for i := 0; i < len(nse.NetworkServiceNames); i++ {
 			ns := nse.NetworkServiceNames[i]
-			var value int32
-			stored, _ := n.nsCounts.LoadOrStore(ns, &value)
-			if nse.ExpirationTime != nil && nse.ExpirationTime.Seconds < 0 {
-				atomic.AddInt32(stored, -1)
-			} else {
-				atomic.AddInt32(stored, 1)
+			state, ok := n.nsStates.Load(ns)
+			if !ok {
+				return
 			}
-			duration := time.Until(nse.ExpirationTime.AsTime())
-			timer := time.AfterFunc(duration, func() {
-				if atomic.AddInt32(stored, -1) <= 0 {
-					if ctx, ok := n.contexts.Load(ns); ok {
+			state.Lock()
+			timer, ok := state.Timers[nse.Name]
+			expirationDuration := time.Until(nse.ExpirationTime.AsTime().Local())
+			if !ok {
+				if expirationDuration > 0 {
+					state.Timers[nse.Name] = time.AfterFunc(expirationDuration, func() {
+						state.Lock()
+						ctx := state.Context
+						delete(state.Timers, nse.Name)
+						state.Unlock()
 						_, _ = n.Unregister(ctx, &registry.NetworkService{Name: ns})
+					})
+					state.Unlock()
+				}
+				continue
+			}
+
+			if expirationDuration < 0 {
+				delete(state.Timers, nse.Name)
+				if timer.Stop() {
+					if len(state.Timers) == 0 {
```

Comment:
I suppose we don't need such check here because we already have it in `Unregister`.

Reply:
It is free to check because we're underlock here.

Reply:
So then we should possibly add the same check to the `time.AfterFunc(...)` to make it look consistent?